torch/_inductor/codegen/cpp_gemm_template.py

```diff
@@ -1046,26 +1079,44 @@ def prep_weight(
             # Require W layout to be fixed & contiguous, happens inplace.
             ir.ExternKernel.require_contiguous(W)
 
-        def _is_int8_gemm(inputs):
-            return (
-                isinstance(inputs[0], ir.IRNode)
-                and inputs[0].get_dtype() in [torch.uint8, torch.int8]
-            ) or (
-                isinstance(inputs[0], torch.Tensor)
-                and inputs[0].dtype in [torch.uint8, torch.int8]
-            )
-
-        if _is_int8_gemm(new_inputs):
+        if not skip_int8_compensation and _is_int8_gemm(new_inputs):
             BCompensate = None
-            if isinstance(W, ir.IRNode):
-                BCompensate = V.graph.add_tensor_constant(
-                    V.graph.constants[W.get_name() + "_BMatrixCompens"],
-                    W.get_name() + "_BMatrixCompens",
-                )
+            x_w_scale = None
+            if use_int8_fast_compensation_path:
+                # new_inputs has been reordered: [x, w, optional[bias], x_scale, x_zp, w_scale, w_zp]
+                x_scale = new_inputs[-4]
+                x_zp = new_inputs[-3]
+                w_scale = new_inputs[-2]
+                if isinstance(W, ir.IRNode):
+                    BCompensate = V.graph.add_tensor_constant(
+                        V.graph.constants[W.get_name() + "_BMatrixCompens"],
+                        W.get_name() + "_BMatrixCompens",
+                    )
```

Comment:
Looks like the code to compute compensate can be shared with the non-fast path. Can we extract this common code?

Reply:
Thanks for the comment and merge these 2 branches.